File: services/feasibility/src/feasibility/telemetry.py
"""OpenTelemetry tracing for the feasibility service.

One export target: the collector. Not Tempo directly — a service should know
one endpoint, so sampling, redaction or a second backend has one place to live.

The consuming half of the trace that starts in tasking-api. What arrives on a
NATS message is a W3C ``traceparent`` naming the outbox publish span; extracting
it is what turns "this service produced some spans" into "this service's work
appears under the HTTP request that asked for it".
"""

from __future__ import annotations

import logging
import os
from typing import TYPE_CHECKING, Any

from opentelemetry import trace
from opentelemetry.baggage.propagation import W3CBaggagePropagator
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.propagate import extract, inject, set_global_textmap
from opentelemetry.propagators.composite import CompositePropagator
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.trace.sampling import ParentBased, TraceIdRatioBased
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

if TYPE_CHECKING:
    from collections.abc import Mapping
    from contextlib import AbstractContextManager

log = logging.getLogger(__name__)

SCOPE_NAME = "feasibility-service"

# Half a second, matching tasking-api. The default is five, which is longer than
# the end-to-end test will wait and longer than a demo's patience for a trace to
# appear.
_EXPORT_INTERVAL_MS = 500


def setup(
    *,
    service_name: str = "feasibility-service",
    service_version: str | None = None,
    environment: str | None = None,
    endpoint: str | None = None,
    sample_ratio: float | None = None,
) -> TracerProvider | None:
    """Install a global tracer provider. Returns None when tracing is off.

    An unreachable collector is NOT a startup failure. Tracing is observability,
    not function; refusing to consume messages because a telemetry backend is
    down would make the observability stack an availability dependency, which is
    exactly backwards. The exporter retries in the background and drops spans in
    the meantime.
    """
    # The propagator is installed even when export is off. It costs nothing, and
    # it means a traceparent still travels THROUGH this service to the next one
    # — so a partially instrumented deployment produces one trace with a gap
    # rather than two unrelated traces.
    #
    # In Python this is belt-and-braces, and that is worth stating because the
    # Go side is not. Measured: opentelemetry-python's DEFAULT global propagator
    # is already a CompositePropagator over traceparent, tracestate and baggage,
    # whereas Go's default is a no-op and the equivalent call there is
    # load-bearing. Setting it here is still right — OTEL_PROPAGATORS can
    # replace the default, and a future release could change it — but no test
    # can distinguish this line from its absence, so none pretends to.
    #
    # In Python this is belt-and-braces, and that is worth stating because the
    # Go side is not. Measured: opentelemetry-python's DEFAULT global propagator
    # is already a CompositePropagator over traceparent, tracestate and baggage,
    # whereas Go's default is a no-op and the equivalent call there is
    # load-bearing. Setting it here is still right — OTEL_PROPAGATORS can
    # replace the default, and a future release could change it — but no test
    # can distinguish this line from its absence, so no test pretends to.
    set_global_textmap(
        CompositePropagator([TraceContextTextMapPropagator(), W3CBaggagePropagator()])
    )

    endpoint = (
        endpoint
        if endpoint is not None
        else os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "otel-collector:4317")
    )
    if not endpoint:
        return None

    if sample_ratio is None:
        sample_ratio = float(os.getenv("TRACE_SAMPLE_RATIO", "1.0"))
    if not 0.0 <= sample_ratio <= 1.0:
        # Rejected rather than clamped. A value outside the range means the
        # author believed something about this knob that is not true, and
        # quietly treating 2.0 as 1.0 leaves the belief in place.
        message = f"TRACE_SAMPLE_RATIO must be between 0 and 1, got {sample_ratio}"
        raise ValueError(message)

    resource = Resource.create(
        {
            "service.name": service_name,
            "service.version": service_version or os.getenv("OVERPASS_VERSION") or "dev",
            "deployment.environment": (environment or os.getenv("OVERPASS_ENV") or "development"),
        }
    )
    provider = TracerProvider(
        resource=resource,
        # ParentBased, so a decision made upstream is honoured rather than
        # re-rolled. Re-rolling per service is how a trace gets holes in it.
        sampler=ParentBased(TraceIdRatioBased(sample_ratio)),
    )
    provider.add_span_processor(
        BatchSpanProcessor(
            # Plaintext: the collector is on the compose network, and TLS
            # between two containers on one bridge buys nothing while adding
            # certificates to a stack whose point is starting in one command.
            OTLPSpanExporter(endpoint=endpoint, insecure=True),
            schedule_delay_millis=_EXPORT_INTERVAL_MS,
        )
    )
    trace.set_tracer_provider(provider)
    return provider


def tracer() -> trace.Tracer:
    """This service's tracer."""
    return trace.get_tracer(SCOPE_NAME)


def consumer_span(
    name: str,
    headers: Mapping[str, str] | None,
    attributes: dict[str, Any] | None = None,
) -> AbstractContextManager[trace.Span]:
    """Start a consumer span that is both a CHILD and a LINK of the producer.

    Both, deliberately, and the reason is the shape of an async hop.

    A pure parent-child relationship says the consumer ran *inside* the
    producer, which is how a synchronous call looks — the publish would appear
    to contain the consume, and the producer's duration would swallow however
    long the message sat in the queue. A pure link says the two are related and
    loses the causal chain: the consumer becomes a root, and asking "what
    happened to this request" returns two traces that a human has to join.

    Child gives the chain. Link states, in the model rather than in a comment,
    that this is a message-driven continuation and not a nested call. Tempo
    renders both, and a reader can see at a glance that the gap between the
    spans is queue time rather than work.

    A message with no traceparent yields a root span rather than an error. That
    is the correct answer for an event published before this instrumentation
    existed, or by a producer that has none.
    """
    parent_context = extract(dict(headers or {}))
    producer = trace.get_current_span(parent_context).get_span_context()

    links = [trace.Link(producer)] if producer.is_valid else []
    return tracer().start_as_current_span(
        name,
        context=parent_context,
        kind=trace.SpanKind.CONSUMER,
        links=links,
        attributes=attributes or {},
    )


def trace_headers(headers: Mapping[str, str] | None = None) -> dict[str, str]:
    """Inject the current span into a header map for the next hop.

    Used by the outbox: an event this service publishes carries the span that
    produced it, exactly as tasking-api's does, so the chain continues past the
    second service instead of ending at it.
    """
    carrier: dict[str, str] = dict(headers or {})
    inject(carrier)
    return carrier


def current_ids() -> tuple[str, str]:
    """Trace and span ids as hex, or empty strings.

    Empty rather than the all-zero id. ``00000000000000000000000000000000`` in a
    log line looks like a real id and matches nothing, which sends whoever reads
    it looking for a trace that cannot exist.
    """
    span_context = trace.get_current_span().get_span_context()
    if not span_context.is_valid:
        return "", ""
    return format(span_context.trace_id, "032x"), format(span_context.span_id, "016x")


class TraceContextFilter(logging.Filter):
    """Adds trace_id and span_id to every record.

    A filter rather than a formatter change, so it applies wherever the record
    is emitted and whatever format is configured. Empty strings when there is no
    span, so the fields always exist and a log pipeline never has to cope with a
    key that sometimes is not there.
    """

    def filter(self, record: logging.LogRecord) -> bool:
        record.trace_id, record.span_id = current_ids()
        return True

File: lib/go/telemetry/telemetry.go
// Package telemetry wires OpenTelemetry tracing and metrics.
//
// One export target: the collector. Not Tempo or Prometheus directly —
// services should know one endpoint, and sampling, redaction or a second
// backend then have one place to live rather than N.
//
// Metrics take the same path as traces, pushed over OTLP rather than scraped
// from a per-service /metrics endpoint. ADR-0018 has the argument; the short
// version is that one of the four services is a worker with no HTTP listener,
// deliberately, and giving it one purely to be scraped is the thing
// docker-compose.yml already refused to do for a readiness probe.
package telemetry

import (
	"context"
	"errors"
	"fmt"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/metric"
	metricnoop "go.opentelemetry.io/otel/metric/noop"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
	tracenoop "go.opentelemetry.io/otel/trace/noop"
)

// ExportInterval is how often the meter provider pushes to the collector.
//
// Ten seconds, matching prometheus.yml's scrape_interval. Faster exports
// samples Prometheus never reads; slower means one scrape sees the same
// cumulative value twice, which flattens a rate into a staircase.
const ExportInterval = 10 * time.Second

// Config is what telemetry needs from the environment.
type Config struct {
	ServiceName    string
	ServiceVersion string
	Environment    string
	// Endpoint is the collector's OTLP gRPC address. Empty disables export.
	Endpoint string
	// SampleRatio is the head sampling ratio for traces. 1 means everything.
	SampleRatio float64
}

// Setup installs global tracer and meter providers and returns one shutdown.
//
// The shutdown must be called and must be given time. Spans are batched and
// metrics are exported on an interval, so a process that exits without
// flushing loses the last few seconds of both — which is reliably the
// interesting few seconds, because it is the ones around whatever made you
// look.
//
// An unreachable collector is NOT a startup failure. Observability is not
// function; refusing to serve traffic because a telemetry backend is down
// would make the observability stack an availability dependency, which is
// exactly backwards. The exporters retry in the background and drop data in
// the meantime.
func Setup(ctx context.Context, cfg Config) (func(context.Context) error, error) {
	// The propagator is installed even when export is off. It costs nothing and
	// it means a traceparent still travels through this service to the next
	// one, so a partially-instrumented deployment produces a trace with a gap
	// rather than two unrelated traces.
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	if cfg.Endpoint == "" {
		otel.SetTracerProvider(tracenoop.NewTracerProvider())
		otel.SetMeterProvider(metricnoop.NewMeterProvider())
		return func(context.Context) error { return nil }, nil
	}

	res, err := buildResource(cfg)
	if err != nil {
		return nil, err
	}

	traceProvider, err := setupTraces(ctx, cfg, res)
	if err != nil {
		return nil, err
	}
	otel.SetTracerProvider(traceProvider)

	meterProvider, err := setupMetrics(ctx, cfg, res)
	if err != nil {
		// Shut the trace provider down rather than leaking its exporter
		// goroutine. Returning an error here aborts startup, so nothing else
		// will ever call the shutdown we would otherwise have returned.
		return nil, errors.Join(err, traceProvider.Shutdown(ctx))
	}
	otel.SetMeterProvider(meterProvider)

	return func(shutdownCtx context.Context) error {
		return errors.Join(
			traceProvider.Shutdown(shutdownCtx),
			meterProvider.Shutdown(shutdownCtx),
		)
	}, nil
}

// buildResource describes this process to both signals.
//
// NewSchemaless, not NewWithAttributes(semconv.SchemaURL, ...).
//
// resource.Merge REFUSES to merge two resources with different schema URLs,
// and resource.Default() carries whichever version the SDK ships. Pinning our
// own semconv version here produced:
//
//	building the resource: conflicting Schema URL:
//	https://opentelemetry.io/schemas/1.43.0 and .../1.26.0
//
// Setup treats that as non-fatal — correctly, telemetry must not stop the
// service — so the service started, logged one warning, and exported nothing.
// It compiled, linted, and passed every unit test, because those call Setup
// with no endpoint and return before this line. Only running it against a real
// collector showed the feature was off.
//
// Schemaless attributes merge with anything. The alternative is chasing the
// SDK's semconv version on every upgrade, and getting that wrong silently
// disables telemetry again.
func buildResource(cfg Config) (*resource.Resource, error) {
	res, err := resource.Merge(resource.Default(), resource.NewSchemaless(
		semconv.ServiceName(cfg.ServiceName),
		semconv.ServiceVersion(cfg.ServiceVersion),
		attribute.String("deployment.environment", cfg.Environment),
	))
	if err != nil {
		return nil, fmt.Errorf("building the resource: %w", err)
	}
	return res, nil
}

func setupTraces(ctx context.Context, cfg Config, res *resource.Resource) (*sdktrace.TracerProvider, error) {
	exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(cfg.Endpoint),
		// Plaintext. The collector is on the compose network; TLS between two
		// containers on the same bridge buys nothing and adds certificates to
		// a stack whose whole point is starting in one command.
		otlptracegrpc.WithInsecure(),
		otlptracegrpc.WithTimeout(5*time.Second),
	)
	if err != nil {
		return nil, fmt.Errorf("creating the OTLP trace exporter: %w", err)
	}

	return sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter,
			// Half a second. The default is five, which is longer than the
			// end-to-end test is willing to wait and longer than a demo's
			// patience for a trace to appear.
			sdktrace.WithBatchTimeout(500*time.Millisecond),
		),
		sdktrace.WithResource(res),
		// ParentBased, so a sampling decision made upstream is honoured rather
		// than re-rolled. Re-rolling per service is how a trace ends up with
		// holes in the middle of it.
		sdktrace.WithSampler(sdktrace.ParentBased(
			sdktrace.TraceIDRatioBased(cfg.SampleRatio),
		)),
	), nil
}

func setupMetrics(ctx context.Context, cfg Config, res *resource.Resource) (*sdkmetric.MeterProvider, error) {
	exporter, err := otlpmetricgrpc.New(ctx,
		otlpmetricgrpc.WithEndpoint(cfg.Endpoint),
		otlpmetricgrpc.WithInsecure(),
		otlpmetricgrpc.WithTimeout(5*time.Second),
	)
	if err != nil {
		return nil, fmt.Errorf("creating the OTLP metric exporter: %w", err)
	}

	// Cumulative temporality, which is the SDK's default for OTLP and what the
	// collector's prometheus exporter expects. Delta would be re-accumulated
	// by the exporter and any collector restart would silently reset every
	// counter on the dashboards.
	return sdkmetric.NewMeterProvider(
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(exporter,
			sdkmetric.WithInterval(ExportInterval),
		)),
		sdkmetric.WithResource(res),
	), nil
}

// Tracer returns a tracer for the given instrumentation scope.
func Tracer(scope string) trace.Tracer { return otel.Tracer(scope) }

// Meter returns a meter for the given instrumentation scope.
func Meter(scope string) metric.Meter { return otel.Meter(scope) }
